File: backend/src/storage/audio-upload-url.service.ts
import { PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

import { s3Client } from '../aws/s3-client';
import { BackendEnv, env } from '../config/env';
import { buildTrackAudioObjectKey } from './audio-object-keys';

type AudioUploadUrlEnv = Pick<
  BackendEnv,
  'S3_AUDIO_BUCKET_NAME' | 'S3_PRESIGNED_UPLOAD_EXPIRES_SECONDS'
>;

export type CreateAudioUploadUrlInput = {
  projectId: string;
  sessionId: string;
  trackId: string;
  contentType: string;
};

export type CreateAudioUploadUrlResult = {
  uploadUrl: string;
  method: 'PUT';
  s3Bucket: string;
  s3Key: string;
  contentType: string;
  expiresInSeconds: number;
};

export type AudioUploadUrlSigner = typeof getSignedUrl;

export class AudioUploadUrlService {
  constructor(
    private readonly client: S3Client,
    private readonly sourceEnv: AudioUploadUrlEnv,
    private readonly signer: AudioUploadUrlSigner = getSignedUrl
  ) {}

  async createUploadUrl(input: CreateAudioUploadUrlInput): Promise<CreateAudioUploadUrlResult> {
    const s3Key = buildTrackAudioObjectKey({
      projectId: input.projectId,
      sessionId: input.sessionId,
      trackId: input.trackId,
    });

    const command = new PutObjectCommand({
      Bucket: this.sourceEnv.S3_AUDIO_BUCKET_NAME,
      Key: s3Key,
      ContentType: input.contentType,
    });

    const uploadUrl = await this.signer(this.client, command, {
      expiresIn: this.sourceEnv.S3_PRESIGNED_UPLOAD_EXPIRES_SECONDS,
    });

    return {
      uploadUrl,
      method: 'PUT',
      s3Bucket: this.sourceEnv.S3_AUDIO_BUCKET_NAME,
      s3Key,
      contentType: input.contentType,
      expiresInSeconds: this.sourceEnv.S3_PRESIGNED_UPLOAD_EXPIRES_SECONDS,
    };
  }
}

export const audioUploadUrlService = new AudioUploadUrlService(s3Client, env);

File: backend-dotnet/src/Loopr.Api/Services/AudioUploadUrlService.cs
using Amazon.S3;
using Amazon.S3.Model;
using Loopr.Api.Configuration;
using Loopr.Api.Storage;
using Microsoft.Extensions.Options;

namespace Loopr.Api.Services;

public sealed record CreateAudioUploadUrlInput(
    string ProjectId,
    string SessionId,
    string TrackId,
    string ContentType
);

public sealed record CreateAudioUploadUrlResult(
    string UploadUrl,
    string Method,
    string S3Bucket,
    string S3Key,
    string ContentType,
    int ExpiresInSeconds
);

public interface IAudioUploadUrlSigner
{
    string CreateUploadUrl(GetPreSignedUrlRequest request);
}

public sealed class S3AudioUploadUrlSigner(IAmazonS3 client) : IAudioUploadUrlSigner
{
    public string CreateUploadUrl(GetPreSignedUrlRequest request)
    {
        return client.GetPreSignedURL(request);
    }
}

public sealed class AudioUploadUrlService(
    IOptions<S3Options> options,
    IAudioUploadUrlSigner signer
)
{
    private const string UploadMethod = "PUT";

    public Task<CreateAudioUploadUrlResult> CreateUploadUrlAsync(
        CreateAudioUploadUrlInput input,
        CancellationToken cancellationToken
    )
    {
        cancellationToken.ThrowIfCancellationRequested();

        var s3Options = options.Value;
        var s3Key = AudioObjectKeys.BuildTrackAudioObjectKey(
            new TrackAudioObjectKeyInput(input.ProjectId, input.SessionId, input.TrackId)
        );
        var request = new GetPreSignedUrlRequest
        {
            BucketName = s3Options.AudioBucketName,
            Key = s3Key,
            ContentType = input.ContentType,
            Verb = HttpVerb.PUT,
            Expires = DateTime.UtcNow.AddSeconds(s3Options.PresignedUploadExpiresSeconds),
        };

        var uploadUrl = signer.CreateUploadUrl(request);

        return Task.FromResult(
            new CreateAudioUploadUrlResult(
                uploadUrl,
                UploadMethod,
                s3Options.AudioBucketName,
                s3Key,
                input.ContentType,
                s3Options.PresignedUploadExpiresSeconds
            )
        );
    }
}
